### ISSUE TITLE
Standard errors with standardized names

### ISSUE THREAD

**Issue by vincentarelbundock** [vincentarelbundock/modelsummary#360]
To the extent possible, standard error types should be named in standardized way across `fixest`, `lfe`, `estimatr` (more?).

In a previous PR, @grantmcdermott pointed to an example with `estimatr` and `fixest`. There were two problems:

1.  The standard error types themselves could have “matching” names.
2.  Rows of the GOF section were not named the same, so the info appears on different rows.

The present issue is meant to discuss Problem 1.

Problem 2 was easy to fix by adding a single line to the `gof_map.R` file. Now we have this:

``` r
library(fixest)
#> fixest 0.9.0, BREAKING changes! (Permanently remove this message with fixest_startup_msg(FALSE).) 
#> - In i():
#>     + the first two arguments have been swapped! Now it's i(factor_var, continuous_var) for interactions. 
#>     + argument 'drop' has been removed (put everything in 'ref' now).
#> - In feglm(): 
#>     + the default family becomes 'gaussian' to be in line with glm(). Hence, for Poisson estimations, please use fepois() instead.
library(estimatr)
library(modelsummary)

mod_feols_hc <- feols(hp ~ mpg, mtcars, vcov = "hetero")
#> Warning: feols(fml = hp ...: vcov is not a valid argument for function feols.
mod_rob_hc <- lm_robust(hp ~ mpg, mtcars, se_type = "stata")

mods_hc <- list("feols HC" = mod_feols_hc, "lm_robust HC" = mod_rob_hc)

msummary(mods_hc,
  output = "markdown",
  gof_omit = "Obs|R2|IC|Lik"
)
```

|             | feols HC | lm\_robust HC |
|:------------|:--------:|:-------------:|
| (Intercept) | 324.082  |    324.082    |
|             | (27.433) |   (28.821)    |
| mpg         |  -8.830  |    -8.830     |
|             | (1.310)  |    (1.256)    |
| Std.Errors  | Standard |      HC1      |

**Comment by vincentarelbundock** [comment]
Standardizing labels probably needs to happy somewhere here:

https://github.com/vincentarelbundock/modelsummary/blob/main/R/get_vcov.R#L192

Maybe we can have mini dictionary that we use for known SE names with the `replace_dict` function.

**Comment by grantmcdermott** [comment]
Agree: A dictionary with standardized names is the way to go. I think the main question is what to use for consistency and interpretability. My first instinct is simply to use the abbreviations employed by sandwich: HC0, HC1, etc. (Clustered SEs and specially named HAC vcovs like NeweyWest would be an exception and treated differently.)

OTOH it might be useful to offer a verbose shorthand like "Heterosked. Rob." that would (I guess) not distinguish between HC0, HC1, etc.

**Comment by vincentarelbundock** [comment]
Sounds good. Most columns are pretty narrow, so I'm not sure the long names would look too good. And people can always supply a named list if they want to control names exactly. So I lean toward:

1. Capitalized `modelsummary` strings whenever supplied explicitly
2. Default `sandwich`-style dictionary for object types we know about

This will require a bit of hand-coding, but if we are mostly thinking about `estimatr`, `fixest` and `lfe`, it's not too bad.

I just spent a bunch of time writing a paper about this package, so my `modelsummary` quota is pretty full right now. But I'm happy to keep exchanging ideas for now.

**Comment by grantmcdermott** [comment]
Cool. Cool. I'm happy to take a stab at this and have the current motivation b/c of teaching / book writing. The only other thing I might (or not) take a look at is returning "Default" instead of "IID" when vcov = NULL. Will submit a PR when I get a chance!

**Comment by vincentarelbundock** [comment]
Excellent!

> returning "Default" instead of "IID" when vcov = NULL.

If we feed the model to `get_vcov_type`, perhaps we can even guess what it is and use our dictionary.

**Comment by grantmcdermott** [comment]
Collecting my thoughts here. To my mind, there are four (five) things that we may want to tackle simultaneously.

1. Consistent SE names, as already discussed above. Basically, a dictionary that references the shorthand strings known to the `vcov` argument (HC1, HC2, etc.)

2. Consistent format for clustered standard errors. For example, at the moment `msummary(fixest::feols(mpg ~ wt, mtcars, vcov = ~ am + vs))` will produce a "Std. Errors" gof entry of `Clustered (am & vs)`. But if we slightly amend to instead pass the vcov argument from modelsummary, i.e. `msummary(fixest::feols(mpg ~ wt, mtcars), vcov = ~ am + vs)`, then the gof entry changes to `C: am + vs`. Easy enough to fix and standardise (and the same things happens for the other models with custom glance methods), but we just need to decide which.

3. While I'm at it, does it make sense for `vcov = "robust"` to fall back to "HC3"? I know that this is the default of `sandwich::vcovHC`. However, `estimatr::lm_robust` defaults to "HC2", while `lfe` and `fixest` default to "HC1" for their heteroskedasticity-consistent SES (following equivalent packages in other languages, most notably Stata). I'm not saying that it makes sense to argue the case from sociological norms &mdash; and the "stata" alias already provides cover for HC1, at least. Nevertheless, it's worth thinking about, given the likely target audience for this functionality and the fact that Stata is where we have likely inherited the "robust" shorthand (which I'm not a fan of TBH b/c its ambiguity, but that ship sailed a long time ago).

4. Inferring the "default" SEs from models when `vcov = NULL`. For standard (g)lm-type models, this will be easy since we know it's IID. For the models with custom glance methods (currently: `fixest`, `lfe`, `estimatr`, and `did`); also easy since we basically handle that already. For other model classes, should we just take the safe option and print "Std. Errors: Default". I must admit that I know less about the default behaviour of `lme4` and some other popular package classes than I might do.

5. P.S. This is tangential to the issue thread, but somewhat related to the previous point so I'll jot it down quickly: Is the current strategy of hardcoding _exceptions_ to the `vcov` argument efficient? Would it maybe be safer / more scaleable to hardcore _inclusions_ (i.e. based on supported inherited model classes like 'lm', 'plm', etc)? OTOH, I suppose many inappropriate methods will through an error and at some point you may just have to accept that a user should know what they're doing...

**Comment by vincentarelbundock** [comment]
Thanks, this is **super** useful.

>     1. Consistent SE names, as already discussed above. Basically, a dictionary that references the shorthand strings known to the `vcov` argument (HC1, HC2, etc.)

With solution to 5, maybe we don't need an actual dict. See below.

>     2. Consistent format for clustered standard errors. For example, at the moment `msummary(fixest::feols(mpg ~ wt, mtcars, vcov = ~ am + vs))` will produce a "Std. Errors" gof entry of `Clustered (am & vs)`. But if we slightly amend to instead pass the vcov argument from modelsummary, i.e. `msummary(fixest::feols(mpg ~ wt, mtcars), vcov = ~ am + vs)`, then the gof entry changes to `C: am + vs`. Easy enough to fix and standardise (and the same thing happens to other model objects with custom glance methods), but we just need to decide which.

Your version of "Clustered (am + vs)" looks better, I think. However, I worry that it might be too "wide" when variable names are longer. This why we picked "C:" format last time, I think. Do you think that's a concern?

>     3. While I'm at it, does it make sense for `vcov = "robust"` to fall back to "HC3"?

I don't have a strong view here and am not aware of a scientific reason to favor one, and I'd be curious to know why `estimatr` and `fixest` picked different. In my mind, the `sandwich` default has always been what robust standard errors *are* -- or at least what the in *R* -- because it's the first thing I ever learned and used. I kind of like it as a default, if only as a nod to 20 years of loyal service by `sandwich`. With the `stata` alias, I'm not sure that matching other results is necessarily a priority. But again, I don't have a strong view on this and am happy to go with whatever you/others prefer.

>     For other model classes, should we just take the safe option and print "Std. Errors: Default".

Agreed. Also, continue to omit the row entirely if non of the model classes are explicitly supported and `msummary(vcov=NULL)`.

>     5. Would it perhaps be safer / more scaleable to hardcore _inclusions_ (i.e. based on supported inherited model classes like 'lm', 'plm', etc)?

Yes to the whitelist approach! maybe we could define a new S3 method called `get_vcov_label`. The `get_vcov_label.default` would return `NULL` immediately. Then we would have `get_vcov_label.plm`, `get_vcov_label.fixest`, etc. As long as these extractor methods produce an output that "matches" the one produced by the `vcov` argument in `modelsummary`, we probably don't need a separate, comprehensive dictionary. Standardization will just occur at extraction time.

**Comment by vincentarelbundock** [comment]
Important: All of this should still be easy to override by supplying a named list to `vcov`.

**Comment by grantmcdermott** [comment]
Great. I probably won't have time to help on the technical implementation today. But having slept on it, here's what I think we should do for 2 and 3.

2. For clustered SE display, I agree: Shorter is better to avoid wide columns. I don't really like `C:` as the  leader, though. How about we follow `fixest::etable`'s lead and display them as `by: am & vs`?

3. `vcov = 'robust'` should simply implement whatever the _object's_ default "robust" vcov is. This is so obvious to me now that I don't why I didn't see it earlier &mdash; probably too fixated on getting the same output &mdash; but let me explain. For some model objects the default robust vcov will yield cluster-robust SEs (e.g. `did` or `fixest` models with clusters specified). For others that yield be HC1 (e.g. `fixest` or `felm` without specified clusters), or HC2 (e.g. `lm_robust`). If the object doesn't have a native robust SE, then we pass it to `sandwich::vcovHC`, which will try to implement HC3 on them. This approach is not to dissimilar to the existing approach, except that we now begin with the exceptions for supported models... basically, all the ones with a custom glance method. Not only will this make the implementation and maintenance much easier on the modelsummary side, but the nice thing is that users should still see an explicit Std. Error type in the table. ("HC2" for `lm_robust` objects, etc.) So it could actually make for a nice pedagogical example, as well.

There's a related case then we should think about too, i.e. `vcov = 'HC'`. Here I'd say that everything holds as per the above paragraph, except that we change from clustered SEs to the (default) heterosk.-consistent ones wherever possible. That's easy enough to do for `fixest` and `felm`, which will then both implement "HC1".

**Comment by vincentarelbundock** [comment]
Green light on 2.

On 3, I'm afraid that returning different types of uncertainty estimates with a single value of the argument (`vcov="robust"`) would violate the [principle of least surprise.](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) Also not currently convinced (but I could be) that we should modify model objects on the fly when users make the explicit decision to pre-specify a vcov when they estimate their model.

Just to make sure I understand the decision tree, did I get all the behaviors right in the cases below?

```{r}
library(fixest)
library(modelsummary)
```

Case 1: Do nothing -- IID 

```r
mod = feols(mpg ~ hp, data = mtcars)
msummary(mod)
```

Case 2: Try a fixest-specific "robust"? How do we know which one is really the "default" robust in each package?

```r
mod = feols(mpg ~ hp, data = mtcars)
msummary(mod, vcov = "robust")
```

Case 3: Return the vcov built in the `fixest` object and use a `get_vcov_label.fixest` method to extract a nice label.

```r
mod = feols(mpg ~ hp, cluster = ~gear, data = mtcars)
msummary(mod)
```

Case 4: raise an error or warning???

```r
mod = feols(mpg ~ hp, cluster = ~gear, data = mtcars)
msummary(mod, vcov = "robust")
```

My main worry is about Case 2. Currently, if users call `vcov="robust"`, they *know* that we will try to extract `HC3` standard errors.

**Comment by vincentarelbundock** [comment]
Another case where I'm not sure what the expected output should be. (Not sure why this produces two GOF rows,. and maybe we need a `"default"` alias)

``` r
library(estimatr)
library(modelsummary)

mod <- list(
  lm_robust(hp ~ mpg, mtcars),
  lm(hp ~ mpg, mtcars)
)

modelsummary(mod, vcov = c("constant", "HC2"), output = "markdown")
```

|             | Model 1  |  Model 2  |
| :---------- | :------: | :-------: |
| (Intercept) | 324.082  |  324.082  |
|             | (29.178) | (29.178)  |
| mpg         | \-8.830  |  \-8.830  |
|             | (1.286)  |  (1.286)  |
| Num.Obs.    |    32    |    32     |
| R2          |  0.602   |   0.602   |
| R2 Adj.     |  0.589   |   0.589   |
| AIC         |          |   336.9   |
| BIC         |          |   341.3   |
| Log.Lik.    |          | \-165.428 |
| F           |          |  45.460   |
| Std.Errors  |   HC2    |    HC2    |
| Std.Errors  | Constant |    HC2    |

**Comment by grantmcdermott** [comment]
> My main worry is about Case 2. Currently, if users call vcov="robust", they know that we will try to extract HC3 standard errors. I think that this predictability is quite nice.

Let me quickly take this one and case 4. 

**Case 2:** I agree about the principle about least surprise. But I think we are _already_ violating it &mdash; in the sense that it either fails or changes the object's own definition of "robust" &mdash; for some of the most common implementation cases. As discussed above, all of `fixest`, `felm`, or `lm_robust` differ from HC3. Even more confusingly, the latter two explicitly call _their_ HCx "robust". For example, the `lfe` package displays its HC1 errors when a user explicitly asks for `robust = TRUE`. 

```r
summary(lfe::felm(mpg ~ wt, mtcars), robust = TRUE)                         
...
Coefficients:
            Estimate Robust s.e t value Pr(>|t|)    
(Intercept)  37.2851     2.1950  16.987  < 2e-16 ***
wt           -5.3445     0.6545  -8.166 4.08e-09 ***
...
```

> How do we know which one is really the "default" robust in each package?

To my mind, there are only four packages/classes that we need to control for and they are all already listed in the `glance_custom.R` file: `fixest`, `felm`, `lm_robust` (`iv_robust`), and `did` (`MP` and `AGGTEobj`). From the perspective of implementing something on the modelsummary side, this makes it easy b/c these objects all have defined methods and/or record their defaults as part of the return object. I'm happy to code something up if you want to see how examples would work in practice before making a call.

**Case 4.** I wouldn't raise an error or warning. Users will see the actual standard error type in the gof section (here: `by: ~ gear`). Again, the issue is that "robust" is ambiguous: it can either mean heterosk-robust or cluster-robust depending on the context and model call. IMO since the user has specified a cluster-correction for this particular model, that's what "robust" means and should fall back too.

If you are tentatively open to seeing an implementation, then I'll code something up and let you see the behaviour and documentation that I have in mind. Naturally, it's still your prerogative to decide against afterwards ;-)

**Comment by vincentarelbundock** [comment]
Yeah, that all sounds good. Frankly, you've thought about these questions more than I have, and you spent more time teaching this stuff, so I'll be very likely to defer to your views on best implementation.

Feel free to start a WIP PR, and we can tweak some things later.

In terms of implementation, I'm not super happy with my own choice to include all this hard-coded stuff in `glance_custom` and `tidy_custom`. I now think those should be reserved as a mechanism for users to customize output.
 
I'm not sure what the better alternative would be. Maybe we can think about making `get_vcov` an S3 method that accepts both the content of the `vcov` argument and the model object, and behaves differently based on the combination of those. That method could then return the actual standard errors and *also* an appropriate label for the `gof` table.

This would "centralize" the vcov logic and make the code clearer, I think.

Let me know if you'd feel more comfortable with me doing this kind of infrastructure work before you look at this.

**Comment by grantmcdermott** [comment]
> Another case where I'm not sure what the expected output should be. (Not sure why this produces two GOF rows,. and maybe we need a "default" alias)

The first `Std. Error` row is correct. There's no way to reliably (or at all?) coerce an `lm_robust` object to a sandwich VCOV, since it doesn't return the iid errors as part of the return.* So, it should throw a warning and print the first `Std. Error.` row.

* As an aside, this is different to `fixest` and `felm`, which both always return iid errors as part of the return object.

**Comment by grantmcdermott** [comment]
Cool. Let me start on WIP PR to handle this then. I think it's best to hold off on any potential move to S3 methods until we have the current hardcoded approach nailed down.

**Comment by grantmcdermott** [comment]
> Not sure why this produces two GOF rows

Just a minor HU to avoid duplicated sleuthing: I've fixed this in the WIP PR.
